### PR TITLE
Adjust Alluxio S3 REST API to conform with S3 API

### PR DESCRIPTION
**Pull request by ZhuTopher**
The following changes are to align the implementation of the Alluxio S3 REST API with expected behaviour for S3 API clients.

---
### Allow CreateBucket calls on existing buckets

Before:
`An error occurred (BucketAlreadyExists) when calling the PutBucket operation: Not allowed to create directory because path already exists: /bucket`

After: 200 OK response

---
### Catch null xAttr when reloading Inodes from journal

This fixes a NPE which would prevent the master from loading a pre-existing journal which had xAttr values on its Inodes.

---
### Re-add exclusion of tagMap field in TaggingData XML

Before:
```
<Tagging>
  <tagMap>
    <tag1>foo</tag1>
    <tag2>bar</tag2>
  </tagMap>
  <TagSet>
    <Tag>
      <Key>tag1</Key>
      <Value>foo</Value>
    </Tag>
    <Tag>
      <Key>tag2</Key>
      <Value>bar</Value>
    </Tag>
  </TagSet>
</Tagging>
```

After:
```
<Tagging>
  <TagSet>
    <Tag>
      <Key>tag1</Key>
      <Value>foo</Value>
    </Tag>
    <Tag>
      <Key>tag2</Key>
      <Value>bar</Value>
    </Tag>
  </TagSet>
</Tagging>
```

---
### Ensure MaxKeys and Prefix are included in ListBucketResult

Before (no prefix in query params):
```
<ListBucketResult>
  <version2>false</version2>
  <IsTruncated>false</IsTruncated>
  <Name>dev</Name>
  <Contents>
    <Key>resources/</Key>
    <Size>0</Size>
    <LastModified>2022-04-12T14:39:43.029Z</LastModified>
  </Contents>
  <MaxKeys>1000</MaxKeys>
  <EncodingType>url</EncodingType>
  <KeyCount>1</KeyCount>
</ListBucketResult>
```

After (no prefix in query params):
```
<ListBucketResult>
  <version2>false</version2>
  <Prefix/>
  <IsTruncated>false</IsTruncated>
  <Name>dev</Name>
  <Contents>
    <Key>resources/</Key>
    <Size>0</Size>
    <LastModified>2022-04-12T14:39:43.029Z</LastModified>
  </Contents>
  <MaxKeys>1000</MaxKeys>
  <EncodingType>url</EncodingType>
  <KeyCount>1</KeyCount>
</ListBucketResult>
```

Before: (max-keys query param set to 0 -> defaulted to 1000)
```
<ListBucketResult>
  <version2>false</version2>
  <IsTruncated>false</IsTruncated>
  <Name>dev</Name>
  <Contents>
    <Key>resources/</Key>
    <Size>0</Size>
    <LastModified>2022-04-12T14:39:43.029Z</LastModified>
  </Contents>
  <MaxKeys>1000</MaxKeys>
  <EncodingType>url</EncodingType>
  <KeyCount>1</KeyCount>
</ListBucketResult>
```

After: (max-keys query param set to 0)
```
<ListBucketResult>
  <version2>false</version2>
  <Prefix/>
  <IsTruncated>false</IsTruncated>
  <Name>dev</Name>
  <MaxKeys>0</MaxKeys>
  <EncodingType>url</EncodingType>
  <KeyCount>0</KeyCount>
</ListBucketResult>
```

---
### Rewrite ListBucketResult methods

Before: (max-keys triggered -> truncated -> NextMarker includes bucket path)
```
<ListBucketResult>
  <version2>false</version2>
  <KeyCount>1</KeyCount>
  <MaxKeys>1</MaxKeys>
  <EncodingType>url</EncodingType>
  <Name>bucket</Name>
  <Contents>
    <Key>file0</Key>
    <Size>0</Size>
    <LastModified>2022-04-14T14:56:05.195Z</LastModified>
  </Contents>
  <NextMarker>/bucket/file0</NextMarker>
  <Prefix/>
  <IsTruncated>true</IsTruncated>
</ListBucketResult>
```

After: (max-keys triggered -> truncated -> NextMarker does not include bucket path)
```
<ListBucketResult>
  <version2>false</version2>
  <KeyCount>1</KeyCount>
  <MaxKeys>1</MaxKeys>
  <EncodingType>url</EncodingType>
  <Name>bucket</Name>
  <Contents>
    <Key>file0</Key>
    <Size>0</Size>
    <LastModified>2022-04-14T14:56:05.195Z</LastModified>
  </Contents>
  <NextMarker>file0</NextMarker>
  <Prefix/>
  <IsTruncated>true</IsTruncated>
</ListBucketResult>
```
---
### Handle edge-cases for 0 or exactly `MaxKeys` results in `ListBucketResult`

Refactored the main body of `ListBucketResult` to handle these edge cases properly.

Before:
- `max-keys=0` -> defaulted to `max-keys=1000`
- `max-keys=${MAX_KEYS}` -> `isTruncated=true` in the response

After:
- `max-keys=0` -> returns a `ListBucketResult` with no keys and `isTruncated=false`
- `max-keys=${MAX_KEYS}` -> `isTruncated=false` in the response

---
### Update implementation for CopyObject to support tagging/metadata directives

Previously there was no support for `x-amz-metadata-directive` nor `x-amz-tagging-directive`. Furthermore there was no copying of metadata/tags when making calls to `CopyObject`. Now there is support for all of the above (see corresponding unit tests).

---
### Fix comments and return error messages to S3 clients

Previously calls to unsupported S3 API actions like `PutBucketAcl`, `PutBucketPolicy`, etc. would get swallowed by the REST handlers for `PutBucket` and then eventually error out or cause unintended behaviour. We now explicitly return an HTTP 500 error code with a message saying we do not support that API action.

### CODE REVIEW

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java
Comment (tcrain) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java:229]: Can you transform the number format exception to an S3 exception here? I think it is an InvalidArgument exception code https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java
Comment (tcrain) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java:168]: Would it still be possible to do these things in a stream instead of a loop? Generally streams are cleaner.

**Comment by HelloHorizon** [comment]
@ZhuTopher do you mind providing more details about what changed before/after you pr? For example, s3 client output, etc.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java
Comment (ZhuTopher) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java:168]: Yes but it involves multiple traversals in a row which I deemed to be not worth the cleanliness of stream processing.

I tried some combinations of filter/map/reduce but none of them were able to process the results in a single traversal. Also using a raw `for`-loop lets me `break` early whereas I couldn't find a way to pre-emptively exit a stream. If you know how please let me know though.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java
Comment (ZhuTopher) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java:229]: That's correct, I'm not sure why the prior implementation was silently converting the result to the default value but that's why I followed suit and added the comment. I'll make the change here then.

**Comment by ZhuTopher** [comment]
@HelloHorizon See the updated PR description

**Review by HelloHorizon — APPROVED** [review]
Thanks @ZhuTopher , remember to update the docs if needed, otherwise LGTM.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java
Comment (tcrain) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java:229]: I mean catch the NumberFormatException and change it to an S3 exception with the correct code. Is that happening somewhere?

**Comment by ZhuTopher** [comment]
Thanks @tcrain for the suggestions & discussion, the `ListBucketResult` implementation has been cleaned up and is using Java streams (again). See the corresponding commits here: https://github.com/ZhuTopher/alluxio/pull/4

@HelloHorizon @tcrain PTAL for a second review, thanks!

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java
Comment (ZhuTopher) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java:194]: I just relied on IntelliJ to suggest how to have a "mutable" String throughout the stream processing. Not sure if this is actually guaranteed to be written to in-order by the stream. @tcrain Any thoughts?

Comment (tcrain) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java:194]: This is fine, it doesn't have to be an atomic reference since this is a sequential stream (see the call to Children.stream) so you also use a reference that is not atomic.

**Comment by ZhuTopher** [comment]
Unrelated test failures:

Java 11 integration tests:
```
Error: 9.023 [ERROR] alluxio.client.cli.job.DistributedCommandsStatsTest.testAsyncPersistCancelStats  Time elapsed: 91.071 s  <<< ERROR!
java.util.concurrent.TimeoutException: Timed out waiting for job 1650481567526 to be one of status [CANCELED] options: WaitForOptions{interval=20, timeout=45000} last value: false
Error: 9.027 [ERROR] alluxio.client.cli.job.DistributedCommandsStatsTest.testAsyncPersistCancelStats  Time elapsed: 90.181 s  <<< ERROR!
java.util.concurrent.TimeoutException: Timed out waiting for job 1650481737914 to be one of status [CANCELED] options: WaitForOptions{interval=20, timeout=45000} last value: false
```

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java
Comment (yuzhu) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: nit， checkNotNull?

Comment (ZhuTopher) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: Outdated implementation

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java
Comment (ZhuTopher) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java:742]: Turns out JAX-RS/Jetty are already handling URL-encoded characters, so there's no need to handle them ourselves. Removing those TODO comments.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java
Comment (tcrain) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java:1270]: With your changes, any delimiter should work right? If the delimiter is / then don't do a recursive listing, otherwise for all other delimiters, do a recursive listing, and it should work?

Comment (ZhuTopher) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java:1270]: That would work, but you _could_ do a "smart" traversal so that you don't need a recursive listing. That's why I left it as a TODO.

eg: with the following file structure
```
/test/delim/
-> foo
-> dir/
     -> img1.png
-> img_store/
     -> img2.png
```

Then you're expected to return the following common prefixes:
```
% aws s3api list-objects-v2 --bucket=czhu-dev --prefix=test/delim --delimiter=/img 
{
    "Contents": [
        {
            "Key": "test/delim/foo",
            "LastModified": "2022-04-20T23:41:19+00:00",
            "ETag": "\"911df44b7ff57801ca8d74568e4ebfbe\"",
            "Size": 27040,
            "StorageClass": "STANDARD"
        }
    ],
    "CommonPrefixes": [
        {
            "Prefix": "test/delim/dir/img"
        },
        {
            "Prefix": "test/delim/img"
        }
    ]
}
```

So ideally you could do a depth-first-traversal on each path and prematurely return whenever you match the delimiter. Maybe _THAT_ should be the TODO as an improvement over naively doing a recursive listing though? Wdyt @yuzhu @tcrain?

Comment (ZhuTopher) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java:1270]: Actually I don't think there's enough time to verify non-"/" delimiters at this point anyway, but yes you're correct that the functionality _is_ actually there if we just did a recursive listing.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java
Comment (ZhuTopher) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java]: Resolving as outdated

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java
Comment (ZhuTopher) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java:194]: Resolving as outdated

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java
Comment (ZhuTopher) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: Resolving as outdated

**Comment by ZhuTopher** [comment]
Unrelated test failures:

Java 8 integration tests:
```
Error: 2.838 [ERROR] alluxio.client.cli.job.DistributedCommandsStatsTest.testAsyncPersistCancelStats  Time elapsed: 91.267 s  <<< ERROR!
java.util.concurrent.TimeoutException: Timed out waiting for job 1650494711060 to be one of status [CANCELED] options: WaitForOptions{interval=20, timeout=45000} last value: false

Error: 2.870 [ERROR] alluxio.client.cli.job.DistributedCommandsStatsTest.testAsyncPersistCancelStats  Time elapsed: 90.163 s  <<< ERROR!
java.util.concurrent.TimeoutException: Timed out waiting for job 1650494881776 to be one of status [CANCELED] options: WaitForOptions{interval=20, timeout=45000} last value: false
```

Java 11 integration tests:
```
Error: 3.636 [ERROR] alluxio.client.fs.FileSystemMasterRestartIntegrationTest.unavailableUfsRecursiveCreate  Time elapsed: 2.971 s  <<< ERROR!
alluxio.exception.AccessControlException: The client user is not authorized so as to be null in server

Error: 3.655 [ERROR] alluxio.client.fs.FileSystemMasterRestartIntegrationTest.unavailableUfsRecursiveCreate  Time elapsed: 2.176 s  <<< ERROR!
alluxio.exception.AccessControlException: The client user is not authorized so as to be null in server
```

**Review by yuzhu — APPROVED** [review]
LGTM, thanks.

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Comment by Xenorith** [comment]
alluxio-bot, cherry-pick this to branch-2.8 please

**Pull request merged**
